### PR TITLE
coprocessor/* : adjust metrics

### PR DESCRIPTION
**Pull request by AndreMouche**
hi,
   This PR split  label `dag` into `select` and `index_scan` in matrics for coprocessor @BusyJay @hicqu PTAL

### CODE REVIEW

File: src/coprocessor/dag/dag.rs [hunk 1]
```diff
@@ -41,28 +41,31 @@ pub struct DAGContext<'s> {
...
         ranges: Vec<KeyRange>,
         snap: &'s Snapshot,
         eval_ctx: Rc<EvalContext>,
-        isolation_level: IsolationLevel,
-        fill_cache: bool,
     ) -> DAGContext<'s> {
         DAGContext {
             req: req,
-            deadline: deadline,
+            deadline: ctx.0,
```
Comment (ngaut) [inline src/coprocessor/dag/dag.rs]: Please do not use tuple, it's hard to maintain.

File: src/coprocessor/select/select.rs [hunk 2]
```diff
@@ -244,7 +244,7 @@ impl<'a> SelectContext<'a> {
         ));
         while row_cnt < self.core.limit {
             if row_cnt & REQUEST_CHECKPOINT == 0 {
-                try!(check_if_outdated(self.deadline, REQ_TYPE_SELECT));
+                try!(check_if_outdated(self.deadline, true));
```
Comment (BusyJay) [inline src/coprocessor/select/select.rs]: Should be false too.

File: src/coprocessor/select/select.rs [hunk 3]
```diff
@@ -208,7 +208,7 @@ impl<'a> SelectContext<'a> {
                 break;
             }
             collected += try!(self.get_idx_row_from_range(r));
-            try!(check_if_outdated(self.deadline, REQ_TYPE_SELECT));
+            try!(check_if_outdated(self.deadline, true));
```
Comment (BusyJay) [inline src/coprocessor/select/select.rs]: I think it should be false.

File: src/coprocessor/endpoint.rs [hunk 4]
```diff
@@ -660,14 +666,6 @@ mod tests {
     use std::time::Duration;
 
     #[test]
-    fn test_get_req_type_str() {
```
Comment (BusyJay) [inline src/coprocessor/endpoint.rs]: Why remove the case?

File: src/coprocessor/endpoint.rs [hunk 5]
```diff
@@ -242,13 +247,19 @@ impl RequestTask {
                     Err(box_err!(e))
                 } else {
                     start_ts = Some(dag.get_start_ts());
+                    if let Some(scan) = dag.get_executors().iter().next() {
```
Comment (BusyJay) [inline src/coprocessor/endpoint.rs]: The first node is always scan?

Comment (AndreMouche) [reply src/coprocessor/endpoint.rs]: yes @BusyJay

File: src/coprocessor/endpoint.rs (hunk 4 above)
Comment (AndreMouche) [reply src/coprocessor/endpoint.rs]: Since function `get_req_type_str ` has been removed @BusyJay

Comment (BusyJay) [reply src/coprocessor/endpoint.rs]: But there is new function named `get_scan_tag`.

File: src/coprocessor/dag/dag.rs [hunk 6]
```diff
@@ -41,28 +41,31 @@ pub struct DAGContext<'s> {
     eval_ctx: Rc<EvalContext>,
     isolation_level: IsolationLevel,
     fill_cache: bool,
+    table_scan: bool,
 }
 
+// deadline, isolation_level, fill_cache, table_scan
+pub type ReqCtx = (Instant, IsolationLevel, bool, bool);
```
Comment (siddontang) [inline src/coprocessor/dag/dag.rs]: use structure, not tuple.

File: src/coprocessor/mod.rs [hunk 7]
```diff
@@ -37,7 +37,7 @@ quick_error! {
             description("key is locked")
             display("locked {:?}", l)
         }
-        Outdated(deadline: Instant, now: Instant, tp: i64) {
+        Outdated(deadline: Instant, now: Instant, tag:&'static str) {
```
Comment (BusyJay) [inline src/coprocessor/mod.rs]: space.

File: src/coprocessor/endpoint.rs [hunk 8]
```diff
@@ -208,16 +207,43 @@ enum CopRequest {
     DAG(DAGRequest),
 }
 
+pub struct ReqCtx {
```
Comment (BusyJay) [inline src/coprocessor/endpoint.rs]: Name it `ReqContext` instead.

**Comment by AndreMouche** [comment]
@siddontang @ngaut @hicqu PTAL

File: src/coprocessor/endpoint.rs [hunk 9]
```diff
@@ -97,22 +95,20 @@ impl CopContext {
...
-            for (cf, details) in this_statistics.details() {
+            if this_statistics.count == 0 {
+                continue;
+            }
+            for (cf, details) in this_statistics.stat.details() {
                 for (tag, count) in details {
                     COPR_SCAN_DETAILS
                         .with_label_values(&[type_str, cf, tag])
-                        .observe(count as f64 / task_count as f64);
+                        .observe(count as f64 / this_statistics.count as f64);
```
Comment (siddontang) [inline src/coprocessor/endpoint.rs]: I don't know exactly what is the metric COPR_SCAN_DETAILS used for, and why do we need to use divide the count here. Can you give me a metric example?

Comment (AndreMouche) [reply src/coprocessor/endpoint.rs]: We will implement `LocalHistogramVec` and use it instead later @siddontang

Comment (siddontang) [reply src/coprocessor/endpoint.rs]: Got it

**Pull request merged**